blk/components/virt.c

```diff
@@ -0,0 +1,345 @@
+    }
+
+    //@ericc: Figure out a better way to assign partitions to clients
+    int client_idx = 0;
+    int num_parts = 0;
+    for (int i = 0; i < MBR_MAX_PRIMARY_PARTITIONS; i++) {
+        if (mbr.partitions[i].type == MBR_PARTITION_TYPE_EMPTY) {
+            continue;
+        } else {
+            num_parts++;
+        }
+        
+        if (client_idx < BLK_NUM_CLIENTS) {
+            clients[client_idx].start_sector = mbr.partitions[i].lba_start;
+            clients[client_idx].sectors = mbr.partitions[i].sectors;
+            client_idx++;
+        }
+
+        if (mbr.partitions[i].lba_start % (BLK_TRANSFER_SIZE / MBR_SECTOR_SIZE) != 0) {
+            LOG_BLK_VIRT_ERR("Partition %d start sector %d not aligned to sDDF transfer size\n", i, mbr.partitions[i].lba_start);
+            return;
+        }
+    }
+
+    if (num_parts < BLK_NUM_CLIENTS) {
+        LOG_BLK_VIRT_ERR("Not enough partitions to assign to clients\n");
+        return;
+    }
+    
+    ((blk_storage_info_t *)blk_config)->sector_size = ((blk_storage_info_t *)blk_config_driver)->sector_size;
+    ((blk_storage_info_t *)blk_config)->block_size = ((blk_storage_info_t *)blk_config_driver)->block_size;
+    ((blk_storage_info_t *)blk_config)->size = clients[0].sectors / (BLK_TRANSFER_SIZE / MBR_SECTOR_SIZE);
+    ((blk_storage_info_t *)blk_config)->read_only = false;
+    ((blk_storage_info_t *)blk_config)->ready = true;
+#if BLK_NUM_CLIENTS > 1
+    ((blk_storage_info_t *)blk_config2)->sector_size = ((blk_storage_info_t *)blk_config_driver)->sector_size;
+    ((blk_storage_info_t *)blk_config2)->block_size = ((blk_storage_info_t *)blk_config_driver)->block_size;
+    ((blk_storage_info_t *)blk_config2)->size = clients[1].sectors / (BLK_TRANSFER_SIZE / MBR_SECTOR_SIZE);
+    ((blk_storage_info_t *)blk_config2)->read_only = false;
+    ((blk_storage_info_t *)blk_config2)->ready = true;
+#endif
+}
+
+static void request_mbr() {
+    uintptr_t mbr_addr;
+    fsmem_alloc(&fsmem_data, &mbr_addr, 1);
+    
+    uint64_t mbr_req_id;
+    ds_data_t mbr_req_data = {0, 0, 0, mbr_addr, 1, 0};
+    datastore_alloc(&ds, &mbr_req_data, &mbr_req_id);
+    
+    blk_enqueue_req(&drv_h, READ_BLOCKS, mbr_addr, 0, 1, mbr_req_id);
+
+    microkit_notify(DRIVER_CH);
+}
+
+static bool handle_mbr_reply() {
+    if (blk_resp_queue_empty(&drv_h)) {
+        LOG_BLK_VIRT_ERR("Trying to request sector 0, no response from driver\n");
+        return false;
+    }
+
+    blk_response_status_t drv_status;
+    uint16_t drv_success_count;
+    uint32_t drv_resp_id;
+    blk_dequeue_resp(&drv_h, &drv_status, &drv_success_count, &drv_resp_id);
+
+    ds_data_t mbr_req_data;
+    datastore_retrieve(&ds, drv_resp_id, &mbr_req_data);
+
+    if (drv_status != SUCCESS) {
+        LOG_BLK_VIRT_ERR("Failed to read sector 0 from driver\n");
+        return false;
+    }
+    
+    microkit_arm_vspace_data_invalidate(mbr_req_data.drv_addr, mbr_req_data.drv_addr + (BLK_TRANSFER_SIZE * mbr_req_data.count));
```

Comment:
This shouldn't be here, it should be done by the driver doing DMA.

Reply:
After reading the sDDF design document, the driver doesn't have a mapping, so it needs to be done on a higher level.

Would be good if Microkit provides infrastructure to do the necessary copying and cleaning/invalidating of data in one function. That way it can also use cache instructions if available instead of syscalls (if RX data is mapped read-only you can just clean & invalidate it instead of invalidating it). This is exactly the kind of tricky code you don't want to let clients or all virtualisers do.

Reply:
I don't think it's for Microkit to be copying data around but doing having one API that automatically uses the user-space caching instructions when possible is planned.